paper-server/patches/sources/net/minecraft/network/Connection.java.patch

```diff
@@ -232,21 +233,38 @@
          }
  
          if (this.tickCount++ % 20 == 0) {
-@@ -432,12 +_,13 @@
+@@ -432,12 +_,20 @@
      }
  
      public void disconnect(DisconnectionDetails disconnectionDetails) {
 +        this.preparing = false; // Spigot
          if (this.channel == null) {
              this.delayedDisconnect = disconnectionDetails;
-         }
+-        }
++            return; // Paper - Defer disconnect until channel is available.
++        }
++
++        // Paper start - Fix PlayerLoginEvent firing multiple times.
++        if (!this.disconnecting.compareAndSet(false, true)) {
```

Comment:
I haven't tested it, but it looks like setting this *before* the isConnected call below means the actual disconnect logic is skipped?